File: lib/oddsApi.ts
import { DEMO_MATCHES } from "@/lib/data";
import type { BookmakerOdds, Market, OddsSnapshot } from "@/lib/types";

interface OddsApiOutcome {
  name: string;
  price: number;
  point?: number;
}

interface OddsApiMarket {
  key: Market;
  last_update: string;
  outcomes: OddsApiOutcome[];
}

interface OddsApiBookmaker {
  key: string;
  title: string;
  last_update: string;
  markets: OddsApiMarket[];
}

interface OddsApiEvent {
  id: string;
  sport_key: string;
  sport_title: string;
  commence_time: string;
  home_team: string;
  away_team: string;
  bookmakers: OddsApiBookmaker[];
}

const ODDS_API_BASE = "https://api.the-odds-api.com/v4";
const SPORT_KEY = "soccer_fifa_world_cup";

function normalizeName(value: string) {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
}

function findMatchId(event: OddsApiEvent) {
  const home = normalizeName(event.home_team);
  const away = normalizeName(event.away_team);
  const commenceDate = event.commence_time.slice(0, 10);

  const exact = DEMO_MATCHES.find((match) => {
    const matchHome = normalizeName(match.homeTeam.name);
    const matchAway = normalizeName(match.awayTeam.name);
    return (
      match.kickoff.slice(0, 10) === commenceDate &&
      ((matchHome === home && matchAway === away) || (matchHome === away && matchAway === home))
    );
  });

  if (exact) return exact.id;

  const fuzzy = DEMO_MATCHES.find((match) => {
    const homeName = normalizeName(match.homeTeam.name);
    const awayName = normalizeName(match.awayTeam.name);
    return (
      (home.includes(homeName) || homeName.includes(home)) &&
      (away.includes(awayName) || awayName.includes(away))
    );
  });

  return fuzzy?.id;
}

function mapBookmakers(bookmakers: OddsApiBookmaker[]): BookmakerOdds[] {
  return bookmakers.map((bookmaker) => ({
    key: bookmaker.key,
    title: bookmaker.title,
    lastUpdate: bookmaker.last_update,
    markets: bookmaker.markets.map((market) => ({
      key: market.key,
      lastUpdate: market.last_update,
      outcomes: market.outcomes.map((outcome) => ({
        name: outcome.name,
        price: outcome.price,
        point: outcome.point
      }))
    }))
  }));
}

export async function fetchWorldCupOdds(): Promise<OddsSnapshot[]> {
  const apiKey = process.env.ODDS_API_KEY;
  if (!apiKey) return [];

  const params = new URLSearchParams({
    apiKey,
    regions: process.env.ODDS_API_REGIONS ?? "us,uk,eu",
    markets: process.env.ODDS_API_MARKETS ?? "h2h,spreads,totals",
    oddsFormat: "decimal",
    dateFormat: "iso"
  });

  const bookmakers = process.env.ODDS_API_BOOKMAKERS;
  if (bookmakers) params.set("bookmakers", bookmakers);

  const response = await fetch(`${ODDS_API_BASE}/sports/${SPORT_KEY}/odds?${params.toString()}`, {
    next: { revalidate: 300 }
  });

  if (!response.ok) {
    return [];
  }

  const events = (await response.json()) as OddsApiEvent[];
  const capturedAt = new Date().toISOString();
  const snapshots: OddsSnapshot[] = [];

  events.forEach((event) => {
    const matchId = findMatchId(event);
    if (!matchId) return;

    snapshots.push({
      matchId,
      source: "the-odds-api",
      stale: false,
      capturedAt,
      bookmakers: mapBookmakers(event.bookmakers)
    });
  });

  return snapshots;
}

File: backend/app/odds_api.py
from __future__ import annotations

import os
import re
from datetime import datetime, timezone

import httpx

from .models import DEMO_MATCHES

ODDS_API_BASE = "https://api.the-odds-api.com/v4"
SPORT_KEY = "soccer_fifa_world_cup"


def normalize_name(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", " ", value.lower()).strip()


def find_match_id(event: dict) -> str | None:
    home = normalize_name(event.get("home_team", ""))
    away = normalize_name(event.get("away_team", ""))
    date = event.get("commence_time", "")[:10]
    for match in DEMO_MATCHES:
        match_home = normalize_name(match.home_team.name)
        match_away = normalize_name(match.away_team.name)
        if match.kickoff[:10] == date and ((home == match_home and away == match_away) or (home == match_away and away == match_home)):
            return match.id
    for match in DEMO_MATCHES:
        match_home = normalize_name(match.home_team.name)
        match_away = normalize_name(match.away_team.name)
        if (match_home in home or home in match_home) and (match_away in away or away in match_away):
            return match.id
    return None


async def fetch_world_cup_odds() -> list[dict]:
    api_key = os.getenv("ODDS_API_KEY")
    if not api_key:
        return []

    params = {
        "apiKey": api_key,
        "regions": os.getenv("ODDS_API_REGIONS", "us,uk,eu"),
        "markets": os.getenv("ODDS_API_MARKETS", "h2h,spreads,totals"),
        "oddsFormat": "decimal",
        "dateFormat": "iso",
    }
    bookmakers = os.getenv("ODDS_API_BOOKMAKERS")
    if bookmakers:
        params["bookmakers"] = bookmakers

    async with httpx.AsyncClient(timeout=12) as client:
        response = await client.get(f"{ODDS_API_BASE}/sports/{SPORT_KEY}/odds", params=params)
    if response.status_code >= 400:
        return []

    captured_at = datetime.now(timezone.utc).isoformat()
    snapshots = []
    for event in response.json():
        match_id = find_match_id(event)
        if not match_id:
            continue
        snapshots.append(
            {
                "matchId": match_id,
                "source": "the-odds-api",
                "stale": False,
                "capturedAt": captured_at,
                "bookmakers": [
                    {
                        "key": bookmaker["key"],
                        "title": bookmaker["title"],
                        "lastUpdate": bookmaker["last_update"],
                        "markets": [
                            {
                                "key": market["key"],
                                "lastUpdate": market["last_update"],
                                "outcomes": [
                                    {
                                        "name": outcome["name"],
                                        "price": outcome["price"],
                                        **({"point": outcome["point"]} if "point" in outcome else {}),
                                    }
                                    for outcome in market["outcomes"]
                                ],
                            }
                            for market in bookmaker["markets"]
                        ],
                    }
                    for bookmaker in event.get("bookmakers", [])
                ],
            }
        )
    return snapshots
